text/0000-default-and-expanded-rustc-errors.md

```diff
@@ -0,0 +1,290 @@
+The proposal is a lighter error format focused on the code the user wrote.  Messages that help understand why an error occurred appear as labels on the source.  The goals of this new format are to:
+
+* Create something that's visually easy to parse
+* Remove noise/unnecessary information
+* Present information in a way that works well for new developers, post-onboarding, and experienced developers without special configuration
+* Draw inspiration from Elm as well as Dybuk and other systems that have already improved on the kind of errors that Rust has.
+
+In order to accomplish this, the proposed design needs to satisfy a number of constraints to make the result maximally flexible across various terminals:
+
+* Multiple errors beside each other should be clearly separate and not muddled together.
+* Each error message should draw the eye to where the error occurs with sufficient context to understand why the error occurs.
+* Each error should have a "header" section that is visually distinct from the code section.
+* Code should visually stand out from text and other error messages.  This allows the developer to immediately recognize their code.
+* Error messages should be just as readable when not using colors (eg for users of black-and-white terminals, color-impaired readers, weird color schemes that we can't predict, or just people that turn colors off)
+* Be careful using “ascii art” and avoid unicode. Instead look for ways to show the information concisely that will work across the broadest number of terminals.  We expect IDEs to possibly allow for a more graphical error in the future.
+* Where possible, use labels on the source itself rather than sentence "notes" at the end.  
+* Keep filename:line easy to spot for people who use editors that let them click on errors
+
+### Header
+
+```
+error[E0499]: cannot borrow `foo.bar1` as mutable more than once at a time
+  --> src/test/compile-fail/borrowck/borrowck-borrow-from-owned-ptr.rs:29:22
+```
+
+The header still serves the original purpose of knowing: a) if it's a warning or error, b) the text of the warning/error, and c) the location of this warning/error.  We keep the error code, now a part of the error indicator, as a way to help improve search results. 
+
+### Line number column
+
+```
+   |
+28 |
+   |
+29 |
+   |
+30 |
+31 |
+   |
+```
+
+The line number column lets you know where the error is occurring in the file.  Because we only show lines that are of interest for the given error/warning, we elide lines if they are not annotated as part of the message (we currently use the heuristic to elide after one un-annotated line).  
+
+Inspired by Dybuk and Elm, the line numbers are separated with a 'wall', a separator formed from pipe('|') characters, to clearly distinguish what is a line number from what is source at a glance.
+
+As the wall also forms a way to visually separate distinct errors, we propose extending this concept to also support span-less notes and hints.  For example:
```

Comment:
There are multiple kinds of notes, the knowledge of which I unfortunately assume in a few places.  Hints, suggestions, notes, etc can all be associated with an error message.  Notes (and the like) which do not have span information need to still be associated with the error.  They're less common now that many notes are becoming labels, but they still exist in a few cases.  Where un-spanned notes are associated with an error, I suggest printing them at the bottom of the error block.

A bit of backstory here.  In a previous iteration of the design (currently in nightly), they are left-aligned and colored.  @alexcrichton showed us some examples at one point where he was getting a lot of error messages in a row, all with unspanned notes.  The result had a negative impact on readability because it was harder to discern each separate error.  In this version of the format, we group unspanned notes and indent them so your eye can see a whole error message as a single group, by skipping the indented line and looking for where the next message starts.